ironfish/src/syncer.ts

```diff
@@ -123,16 +162,90 @@ export class Syncer {
       .getConnectedPeers()
       .filter((peer) => peer.features?.syncing && peer.work && peer.work > head.work)
 
-    // Get a random peer with higher work. We do this to encourage
-    // peer diversity so the highest work peer isn't overwhelmed
-    // as well as helping to make sure we don't get stuck with unstable peers
-    if (peers.length > 0) {
-      const peer = ArrayUtils.sampleOrThrow(peers)
+    if (peers.length === 0) {
+      return
+    }
+
+    // If there is only one valid peer to sync from, there is no point in
+    // measuring the connection so begin syncing immediately
+    if (peers.length === 1) {
+      this.startSync(peers[0])
+      return
+    }
+
+    const syncCandidates = ArrayUtils.shuffle(peers).slice(0, 8)
+
+    // If we have been syncing from a peer, we want to include this peer in the
+    // measurement. This will allow us to maintain a connection to a strong peer
+    // if we have found one.
+    if (
+      currentPeerIdentity &&
+      !syncCandidates.find((p) => p.getIdentity() === currentPeerIdentity)
+    ) {
+      const currentPeer = this.peerNetwork.peerManager.getPeer(currentPeerIdentity)
+      if (currentPeer && currentPeer.state.type === 'CONNECTED') {
+        syncCandidates.push(currentPeer)
+      }
+    }
+
+    const peerRtt = new Map<Identity, number>()
+
+    // Measure how long it takes to fetch the genesis block header from each
+    // peer as an estimate of connection quality
+    for (const peer of syncCandidates) {
```

Comment:
I think there's a potential case here where all of the 8 `syncCandidates` could fail to respond to the blockHeader request, and then we end up unable to start syncing even though we may have more peers we could try measuring. Have you thought about continuing to measure peers until we've tried all peers or have measured 8 successfully?

This would have the tradeoff of increasing time spent measuring

Reply:
That's a good point. While it does seem statistically unlikely unless you have <8 peers to begin with, it is possible and it would lead to a stuck state until the next measurement time.

There's 2 potential approaches here that I'm thinking of
1. We wrap these 2 lines https://github.com/iron-fish/ironfish/pull/4371/files#diff-fde96c520c62d9d6f4af77d595f4b9775f195d846d3486a939085fe685a643d9R123-R124 in an `if (this.loader) { ... }` block. This would mean that it would try 8 more peers on the next event-loop execution
2. What you've suggested, where instead of just slicing 8 peers off the top, we just iterate until we've measured 8 peers, or exhausted all candidates. There's still the potential that we don't get a single valid response in the situation of, for example, you only having 2-3 peers and they all time out. In this case, we'll likely still need #1. 

Thinking through this, your suggestion does make it feel more robust, so I'll do that along with the check that we did indeed find a peer

Reply:
Sounds good, I think both 1 and 2 you proposed here make sense to me! 👍